File: api/schemas/topoiq.py
"""Pydantic schemas for TopoIQ API."""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


class TopoIQAnalyzeRequest(BaseModel):
    project_name: str = Field(default="Terrain analysis", max_length=200)
    country: str = Field(default="", max_length=120)
    land_use: Literal["Standard", "Agri-PV"] = "Standard"
    polygons: List[List[Any]] = Field(
        ...,
        description="List of polygon rings. Points can be {lat, lon} or [lon, lat].",
    )
    grid_m: float = Field(default=5.0, gt=0)
    allow_coarsen: bool = False
    contour_minor: float = Field(default=0.5, gt=0)
    contour_major: float = Field(default=1.0, gt=0)
    mask_geojson: Optional[Dict[str, Any]] = Field(
        default=None,
        description="Optional SiteIQ buildable-area GeoJSON to clip terrain to buildable land.",
    )


class TopoIQVerdict(BaseModel):
    label: str
    detail: str


class TopoIQAnalyzeResponse(BaseModel):
    project_name: str
    country: str
    land_use: str
    area_ha: float
    grid_m_requested: float
    grid_m_used: float
    grid_points: int
    dem_zoom: int
    tile_count: int
    terrain_source_used: str
    terrain_source: Dict[str, Any]
    elevation: Dict[str, float]
    slope: Dict[str, Any]
    extras: Dict[str, Any]
    verdict_fixed: TopoIQVerdict
    verdict_tracker: TopoIQVerdict
    terrain_drivers: Dict[str, Any]
    contour_minor: float
    contour_major: float
    disclaimer: str
    bbox: Dict[str, float]
    route_note: Optional[str] = None

File: frontend/src/types/topoiq.ts
export interface TopoIQAnalyzeRequest {
  project_name: string;
  country: string;
  land_use: "Standard" | "Agri-PV";
  polygons: Array<Array<{ lat: number; lon: number }>>;
  grid_m?: number;
  allow_coarsen?: boolean;
  contour_minor?: number;
  contour_major?: number;
  mask_geojson?: unknown;
}

export type TerrainDriverKind = "positive" | "warn" | "neutral";

export interface TerrainDrivers {
  terrain_score: number;
  terrain_score_label: string;
  drivers: Array<[string, string, TerrainDriverKind]>;
  why_bullets: Array<[TerrainDriverKind, string]>;
}

export interface TopoIQAnalyzeResponse {
  project_name: string;
  country: string;
  land_use: string;
  area_ha: number;
  grid_m_requested: number;
  grid_m_used: number;
  grid_points: number;
  dem_zoom: number;
  tile_count: number;
  terrain_source_used: string;
  terrain_source: Record<string, unknown>;
  elevation: {
    z_min: number;
    z_max: number;
    z_range: number;
    center_elev: number;
  };
  slope: {
    mean: number;
    max: number;
    pct_over5: number;
    pct_over10: number;
    bins?: number[];
  };
  extras: Record<string, unknown>;
  verdict_fixed: { label: string; detail: string };
  verdict_tracker: { label: string; detail: string };
  terrain_drivers: TerrainDrivers;
  contour_minor: number;
  contour_major: number;
  disclaimer: string;
  bbox: { lat_c: number; lon_c: number };
  route_note?: string | null;
}

export interface YieldIQAnalyzeRequest {
  lat: number;
  lon: number;
  mount_type: string;
  gcr_1p?: number;
  gcr_2p?: number;
  soiling_loss?: number;
  other_loss?: number;
}

export interface YieldConfig {
  display_name: string;
  spec_y: number;
  pr?: number | null;
  gcr: number;
  shading: number;
  total_loss: number;
  monthly: number[];
  [key: string]: unknown;
}

export interface YieldIQAnalyzeResponse {
  lat: number;
  lon: number;
  mount_type: string;
  raddatabase?: string | null;
  configs: Record<string, YieldConfig>;
  cross_ref_bundle: Record<string, number | null>;
  disclosure: string;
}
